envoy: Better error logging.

jrajahalme opened the pull request:
While we don't want to log too much when debug is not enabled, we
should log all messages on the xDS stream if we ever receive a NACK,
which indicates a bug. To do this keep track if the last message from
Envoy was an ACK or NACK and log sent resources at Info level if we
received a NACK. This helps in cases where we repeatedly send
incorrect configuration to Envoy.

Envoy xDS messages have added an optional Status message, which can
carry an error message from Envoy to Cilium. Log that whenever
included in a request.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/4002)
<!-- Reviewable:end -->

rlenglet commented on pkg/envoy/xds/server.go:
```diff
@@ -178,6 +184,11 @@ type perTypeStreamState struct {
 	// response to a request for this resource type.
 	resourceNames []string
+
+	// nackReceived it true if this stream has ever received a NACK. A NACK indicates
```
This is not consistent with the implementation. If we ever receive an ACK, this is set to `false`. Document that here.

jrajahalme replied:
Thanks for noticing, I changed my mind while I was writing this comment :-)

rlenglet commented on pkg/envoy/xds/server.go:
```diff
@@ -178,6 +184,11 @@ type perTypeStreamState struct {
+
+	// nackReceived it true if this stream has ever received a NACK. A NACK indicates
+	// an error on the xDS protocol, which always is a bug. Further messages on the stream
+	// are then logged at warning level until we receive an ACK.
```
warning -> Info

jrajahalme replied:
fixed.

rlenglet reviewed (approved):
The implementation LGTM. Please fix the comment on the `nackReceived` field.

jrajahalme committed:
envoy: Better error logging.

While we don't want to log too much when debug is not enabled, we
should log all messages on the xDS stream if we ever receive a NACK,
which indicates a bug. To do this keep track if the last message from
Envoy was an ACK or NACK and log sent resources at Info level if we
received a NACK. This helps in cases where we repeatedly send
incorrect configuration to Envoy.

Envoy xDS messages have added an optional Status message, which can
carry an error message from Envoy to Cilium. Log that whenever
included in a request.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>
